### ISSUE TITLE
Crash when signing a document with PADES

### ISSUE THREAD

**Issue by Leuca** [Leuca/CIE-Middleware-RPM-Package#1]
Caused by using the version of PoDoFo shipped by Fedora/RHEL. A patch allows the code to link against a shared version of the library but it is not enough to support PADES.

**Comment by Leuca** [comment]
The culprit seems to be PoDoFo::PdfPagesTree::GetTotalNumberOfPages()
```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f205099edb4, pid=620852, tid=620940
#
# JRE version: OpenJDK Runtime Environment (Red_Hat-11.0.24.0.8-2) (11.0.24+8) (build 11.0.24+8-LTS)
# Java VM: OpenJDK 64-Bit Server VM (Red_Hat-11.0.24.0.8-2) (11.0.24+8-LTS, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libcie-pkcs11.so+0x19edb4]  PoDoFo::PdfPagesTree::GetTotalNumberOfPages() const+0x4
```

**Comment by Leuca** [comment]
After some changes that should improve the broken part of library there is a new problematic frame:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f7e2d39a5e3, pid=1595317, tid=1595814
#
# JRE version: OpenJDK Runtime Environment (Red_Hat-11.0.24.0.8-2) (11.0.24+8) (build 11.0.24+8-LTS)
# Java VM: OpenJDK 64-Bit Server VM (Red_Hat-11.0.24.0.8-2) (11.0.24+8-LTS, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [libcie-pkcs11.so+0x19a5e3]  PoDoFo::PdfImage::LoadFromPngHandle(PoDoFo::PdfFileInputStream*)+0x2d3
```

Note: This crash happens when signing with graphical signature

**Comment by Leuca** [comment]
Also this warning could be interesting. Happens both with and without graphical signature.


```
WARNING: There are more objects (216) in this XRef table than specified in the size key of the trailer directory (200)!
<</ID[<90D904BA435840C0B691035C31575EE3><4C8BCBA130474F1AB7F016C0D03DB451>]/Info 199 0 R/Root 201 0 R/Size 200>>
PDFLENGTH9300166STOPGetNumberOfSignaturestrailer okCatalogref okcatalog okacroform ok 1acroform ok 2acroform ok 3fieldsValue okfieldsValue ok 2fieldsValue is arrayOK nSigns: 0OK m_pPdfDocument
```

**Comment by Leuca** [comment]
Since release 30 the app should be able to sign with PADES without crashing. The signature produced is probably malformed, but I wasn't able to verify it just yet.
The graphical signature doesn't show up on the output pdf.

Slight improvements, but there's still work to be done.

**Comment by Leuca** [comment]
> The culprit seems to be PoDoFo::PdfPagesTree::GetTotalNumberOfPages()
> 
> ```
> # A fatal error has been detected by the Java Runtime Environment:
> #
> #  SIGSEGV (0xb) at pc=0x00007f205099edb4, pid=620852, tid=620940
> #
> # JRE version: OpenJDK Runtime Environment (Red_Hat-11.0.24.0.8-2) (11.0.24+8) (build 11.0.24+8-LTS)
> # Java VM: OpenJDK 64-Bit Server VM (Red_Hat-11.0.24.0.8-2) (11.0.24+8-LTS, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
> # Problematic frame:
> # C  [libcie-pkcs11.so+0x19edb4]  PoDoFo::PdfPagesTree::GetTotalNumberOfPages() const+0x4
> ```

This was because I made the code use a PdfMemDocument before it was initialized the document buffer



> After some changes that should improve the broken part of library there is a new problematic frame:
> 
> ```
> #
> # A fatal error has been detected by the Java Runtime Environment:
> #
> #  SIGSEGV (0xb) at pc=0x00007f7e2d39a5e3, pid=1595317, tid=1595814
> #
> # JRE version: OpenJDK Runtime Environment (Red_Hat-11.0.24.0.8-2) (11.0.24+8) (build 11.0.24+8-LTS)
> # Java VM: OpenJDK 64-Bit Server VM (Red_Hat-11.0.24.0.8-2) (11.0.24+8-LTS, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
> # Problematic frame:
> # C  [libcie-pkcs11.so+0x19a5e3]  PoDoFo::PdfImage::LoadFromPngHandle(PoDoFo::PdfFileInputStream*)+0x2d3
> ```
> 
> Note: This crash happens when signing with graphical signature

This was because the code I wrote didn't handle the signature image correctly



> Also this warning could be interesting. Happens both with and without graphical signature.
> 
> ```
> WARNING: There are more objects (216) in this XRef table than specified in the size key of the trailer directory (200)!
> <</ID[<90D904BA435840C0B691035C31575EE3><4C8BCBA130474F1AB7F016C0D03DB451>]/Info 199 0 R/Root 201 0 R/Size 200>>
> PDFLENGTH9300166STOPGetNumberOfSignaturestrailer okCatalogref okcatalog okacroform ok 1acroform ok 2acroform ok 3fieldsValue okfieldsValue ok 2fieldsValue is arrayOK nSigns: 0OK m_pPdfDocument
> ```

This is harmless, probably because of the source pdf

> Since release 30 the app should be able to sign with PADES without crashing. The signature produced is probably malformed, but I wasn't able to verify it just yet. The graphical signature doesn't show up on the output pdf.
> 
> Slight improvements, but there's still work to be done.

Turns out that upstream uses a fork of PoDoFo and doesn't publish the sources. Looking at the code looks like upstream modified the parts of code that handle making the PADES signature. Hence the PADES signature implementation is closed-source. A new open-source implementation is required.

**Comment by Leuca** [comment]
Fixed in ab47a78
